### PR TITLE
Separate address hashing

### PR DESCRIPTION
**Pull request by nc6**
Draft only at the moment for looking-at-changes purposes.

### CODE REVIEW

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Keys.hs [hunk 1]
```diff
@@ -99,36 +130,79 @@ import Shelley.Spec.Ledger.Crypto
...
+  Genesis :: KeyRole RegularHash
+  GenesisDelegate :: KeyRole RegularHash
+  Payment :: KeyRole AddrHash
+  Staking :: KeyRole AddrHash
+  StakePool :: KeyRole RegularHash
+  -- A block issuer might be either a genesis delegate or a stake pool
+  BlockIssuer :: KeyRole RegularHash
+  -- Witness keys may have either hash type.
+  AWitness :: KeyRole AddrHash
+  RWitness :: KeyRole RegularHash
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Keys.hs:143]: oh my, this is gorgeous! :old_key:

**Review by JaredCorduan — COMMENTED** [review]
this is wonderful @nc6, I love it

File: shelley/chain-and-ledger/executable-spec/cddl-files/shelley.cddl [hunk 2]
```diff
@@ -186,8 +186,9 @@ protocol_param_update =
 
 transaction_witness_set =
   { ?0 => [* vkeywitness ]
-  , ?1 => [* multisig_script ]
-  , ?2 => [* bootstrap_witness ]
+  , ?1 => [* vkeywitness ]
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/cddl-files/shelley.cddl]: Perhaps one of these could be renamed to address_vkey_witness

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Keys.hs [hunk 3]
```diff
@@ -99,36 +130,79 @@ import Shelley.Spec.Ledger.Crypto
...
+
+instance Crypto crypto => IsKeyRole Staking crypto
+
+instance Crypto crypto => IsKeyRole StakePool crypto
+
+instance Crypto crypto => IsKeyRole BlockIssuer crypto
+
+instance Crypto crypto => IsKeyRole AWitness crypto
+
+instance Crypto crypto => IsKeyRole RWitness crypto
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Keys.hs:179]: There must be a way to do this

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Keys.hs [hunk 4]
```diff
@@ -199,32 +279,54 @@ verifySignedDSIGN (VKey vk) vd sigDSIGN =
...
   deriving newtype (NFData, NoUnexpectedThunks, Generic)
 
-deriving instance (Crypto crypto, Typeable disc) => ToCBOR (KeyHash disc crypto)
+deriving instance
+  ( Crypto crypto,
+    Typeable disc,
+    Typeable (KeyRoleHashType disc),
+    Hash.HashAlgorithm (AlgorithmForHashType crypto (KeyRoleHashType disc))
+  ) =>
+  ToCBOR (KeyHash disc crypto)
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Keys.hs:289]: IsKeyRole

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Genesis.hs [hunk 5]
```diff
@@ -212,10 +212,10 @@ initialFundsPseudoTxIn addr =
       error $
         "Unsupported Byron address in the genesis UTxO: " <> show byronAddr
   where
-    pseudoTxIn :: Crypto.Hash (HASH c) a -> TxIn c
+    pseudoTxIn :: Crypto.Hash x a -> TxIn c
     pseudoTxIn h = TxIn (pseudoTxId h) 0
-    pseudoTxId :: Crypto.Hash (HASH c) a -> TxId c
+    pseudoTxId :: Crypto.Hash x a -> TxId c
     pseudoTxId = TxId . castHash
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Genesis.hs]: This looks suspicious. What happens if we have the wrong length?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs [hunk 6]
```diff
@@ -757,13 +809,12 @@ witsVKeyNeeded utxo' tx@(Tx txbody _ _) _genDelegs =
         [ ((Set.map asWitness) . _poolOwners) pool
           | DCertPool (RegPool pool) <- toList $ _certs txbody
         ]
-    certAuthors = Set.map asWitness . Set.fromList $ foldl (++) [] $ fmap getCertHK certificates
-    getCertHK = cwitness
+    certAuthors = foldl' (<>) ([], []) $ fmap cwitness certificates
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs:812]: Accidentally quadratic?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxow.hs [hunk 7]
```diff
@@ -229,8 +236,11 @@ utxoWitnessed =
       verifiedWits tx ?!: InvalidWitnessesUTXOW
 
       let needed = witsVKeyNeeded utxo tx genDelegs
-          missingWitnesses = needed `Set.difference` witsKeyHashes
-          haveNeededWitnesses = case missingWitnesses == Set.empty of
+          missingWitnesses =
+            WitHashes
+              (addrWitHashes needed `Set.difference` addrWitHashes witsKeyHashes)
+              (regWitHashes needed `Set.difference` regWitHashes witsKeyHashes)
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxow.hs:242]: diffWitHashes

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Serialization.hs [hunk 8]
```diff
@@ -97,6 +97,7 @@ import Data.Text (Text)
 import qualified Data.Text as Text
 import Data.Typeable
 import Network.Socket (HostAddress6)
+import Prelude
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Serialization.hs:100]: This is mysterious

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Address.hs [hunk 9]
```diff
@@ -262,10 +272,28 @@ genAddr =
...
-genCredential :: Gen (C.Credential kr)
+genCredential ::
+  forall kr.
+  ( HashAlgorithm
+      ( AlgorithmForHashType
+          OtherCrypto
+          ( KeyRoleHashType kr
+          )
+      )
+  ) =>
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Address.hs:283]: IsKeyRole

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Address.hs [hunk 10]
```diff
@@ -262,10 +272,28 @@ genAddr =
...
-genKeyHash :: Gen (C.Credential kr)
+genKeyHash ::
+  forall kr.
+  ( HashAlgorithm
+      ( AlgorithmForHashType
+          OtherCrypto
+          ( KeyRoleHashType kr
+          )
+      )
+  ) =>
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Address.hs:295]: IsKeyRole

File: shelley/chain-and-ledger/executable-spec/test/shelley-spec-ledger-test.cabal [hunk 11]
```diff
@@ -53,6 +54,7 @@ library
   build-depends:
     aeson,
     base,
+    base16-bytestring,
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/test/shelley-spec-ledger-test.cabal:57]: Is this used?

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Generator/Genesis.hs [hunk 12]
```diff
@@ -150,8 +162,14 @@ genSeed :: Int -> Gen Seed
...
 genKeyHash ::
-  Crypto c =>
-  Gen (KeyHash krole c)
+  ( Crypto c,
+    Hash.HashAlgorithm
+      ( AlgorithmForHashType
+          c
+          (KeyRoleHashType disc)
+      )
+  ) =>
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Generator/Genesis.hs:164]: IsKeyRole

File: shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/NonTraceProperties/Generator.hs [hunk 13]
```diff
@@ -391,13 +392,13 @@ genDCertDeRegKey keys =
 genDCertRetirePool :: KeyPairs -> EpochNo -> Gen DCert
 genDCertRetirePool keys epoch = do
   key <- getAnyStakeKey keys
-  pure $ DCertPool $ RetirePool (coerceKeyRole $ hashKey key) epoch
+  pure $ DCertPool $ RetirePool (unsafeCoerce $ hashKey key) epoch
```
Comment (redxaxder) [inline shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/NonTraceProperties/Generator.hs:395]: Perhaps `coerce` applies here.

File: shelley/chain-and-ledger/executable-spec/test/shelley-spec-ledger-test.cabal (hunk 11 above)
Comment (nc6) [reply shelley/chain-and-ledger/executable-spec/test/shelley-spec-ledger-test.cabal:57]: Turns out it's needed by the Fees module

**Review by redxaxder — APPROVED** [review]

**Review by JaredCorduan — APPROVED** [review]
:+1:

**Comment by nc6** [comment]
bors merge

**Pull request merged**